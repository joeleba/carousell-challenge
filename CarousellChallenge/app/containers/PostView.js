import React, { Component } from 'react';
import { AppRegistry, Text, FlatList, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import { addPost } from '../actions';
import NewContentForm from '../components/NewContentForm';
import CommentThread from '../containers/CommentThread';
import _ from 'lodash'

class PostListView extends Component {
  _keyExtractor(item, index) {
    return item.id;
  }

  render() {
    let post = _.find(
      this.props.posts,
      { id: this.props.navigation.state.params.postId }
    );
    let childrenObjects = _.map(
      post.children, (id) => { return this.props.comments[id] }
    );
    console.log(childrenObjects);
    return (
      <View style={styles.container}>
        <View style={styles.post}>
          <Text> {post.title} </Text>
          <Text> {post.content} </Text>
          <Text
            style={styles.replyButton}
            onPress={() => this.props.dispatchNavigateToNewReply(post.id, post)}>
            Reply
          </Text>
        </View>

        { childrenObjects.length !== 0 &&
          <FlatList
            data={childrenObjects}
            keyExtractor={this._keyExtractor}
            renderItem={({item}) =>
              <View style={styles.comment}>
                <CommentThread comment={item}/>
              </View>
            }
          />
        }
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
   flex: 1,
   paddingTop: 22
  },
  post: {
    padding: 10,
    borderColor: 'powderblue',
    borderBottomWidth: 1,
  },
  comment: {
    padding: 10,
  },
  replyButton: {

  }
})

function mapStateToProps(state) {
  return {
    posts: state.contentReducers.posts,
    comments: state.contentReducers.comments
  }
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToNewReply: (postId, replyingToObject) =>
      dispatch(NavigationActions.navigate({
        routeName: 'NewComment',
        params: {
          postId: postId,
          replyingToObject: replyingToObject
        }
      }))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PostListView)
