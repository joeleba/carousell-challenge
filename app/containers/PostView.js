import React, { Component } from 'react';
import { Text, FlatList, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import Icon from '@expo/vector-icons/FontAwesome';
import _ from 'lodash';

import NewContentForm from '../components/NewContentForm';
import CommentThreadWrapper from '../containers/CommentThread';
import OptionsRow from '../containers/OptionsRow';
import { POST_TYPE } from '../actions';

class PostView extends Component {
  keyExtractor(item, index) {
    return item.id;
  }

  render() {
    const post = this.props.posts[this.props.navigation.state.params.postId];
    const childrenObjects = _.map(post.children, id => this.props.comments[id]);

    return (
      <View style={styles.container}>
        <View style={styles.post}>
          <Icon
            name="close"
            size={20}
            onPress={() => this.props.dispatchNavigateBackToPostListView()}
          />

          <Text style={styles.title}> {post.title} </Text>
          <Text> {post.content} </Text>

          <OptionsRow currentComment={post} postId={post.id} contentType={POST_TYPE} />
        </View>

        { childrenObjects.length !== 0 &&
          <FlatList
            data={childrenObjects}
            keyExtractor={this.keyExtractor}
            renderItem={({ item }) =>
              <View style={styles.comment}>
                <CommentThreadWrapper comment={item} postId={post.id} />
              </View>
            }
          />
        }
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 22,
  },
  post: {
    padding: 10,
    borderColor: 'lightgrey',
    borderBottomWidth: 1,
  },
  comment: {
    padding: 10,
  },
  title: {
    fontSize: 25,
    paddingTop: 10,
  },
});

function mapStateToProps(state) {
  return {
    posts: state.contentReducers.posts,
    comments: state.contentReducers.comments,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateBackToPostListView: () =>
      dispatch(NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'PostListView' })],
      })),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PostView);
