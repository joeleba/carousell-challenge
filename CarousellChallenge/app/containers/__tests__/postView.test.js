import React from 'react';
import { Text, FlatList, StyleSheet, View } from 'react-native';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import PostView from '../PostView';

const mockStore = configureStore();

const stateNestedComments = {
  contentReducers: {
    currentId: 6,
    posts: {
      0: {
        children: [1, 2],
        content: "ccc",
        downvoteCount: 0,
        upvoteCount: 0,
        title: "sad",
        id: 0
      }
    },
    comments: {
      1: {
        children: [3, 4, 5],
        content: "ccc1",
        downvoteCount: 0,
        upvoteCount: 0,
        id: 1
      },
      2: {
        children: [],
        content: "ccc2",
        downvoteCount: 0,
        upvoteCount: 0,
        id: 2
      },
      3: {
        children: [],
        content: "ccc3",
        downvoteCount: 0,
        upvoteCount: 0,
        id: 3
      },
      4: {
        children: [],
        content: "ccc4",
        downvoteCount: 0,
        upvoteCount: 0,
        id: 4
      },
      5: {
        children: [],
        content: "ccc5",
        downvoteCount: 0,
        upvoteCount: 0,
        id: 5
      },
    },
  },
};

const stateNoComment = {
  contentReducers: {
    currentId: 1,
    posts: {
      0: {
        children: [],
        content: "ccc",
        downvoteCount: 0,
        upvoteCount: 0,
        title: "sad",
        id: 0
      }
    },
    comments: {},
  },
};

const navigationProp = {
  state: {
    params: {
      postId: 0
    }
  }
};

describe('Component: PostView', () => {
  it('renders with nested comments', () => {
    const wrapper = shallow(
      <PostView navigation={navigationProp}/>,
      { context: { store: mockStore(stateNestedComments) } },
    );
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.prop('posts')).toEqual(stateNestedComments['contentReducers']['posts']);
    expect(wrapper.prop('comments')).toEqual(stateNestedComments['contentReducers']['comments']);
    expect(wrapper.dive().find(FlatList)).toHaveLength(1);
  });

  it('renders with no comment', () => {
    const wrapper = shallow(
      <PostView navigation={navigationProp}/>,
      { context: { store: mockStore(stateNoComment) } },
    );
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.prop('posts')).toEqual(stateNoComment['contentReducers']['posts']);
    expect(wrapper.prop('comments')).toEqual(stateNoComment['contentReducers']['comments']);
    expect(wrapper.dive().find(FlatList)).toHaveLength(0);
  });
});
