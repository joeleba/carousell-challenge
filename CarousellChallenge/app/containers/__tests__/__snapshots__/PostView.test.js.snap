// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PostView renders with nested comments 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 22,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "lightgrey",
        "padding": 10,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      name="close"
      onPress={[Function]}
      size={20}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 25,
          "paddingTop": 10,
        }
      }
    >
       
      sad
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       
      ccc
       
    </Text>
    <Connect(OptionsRow)
      currentComment={
        Object {
          "children": Array [
            1,
            2,
          ],
          "content": "ccc",
          "downvoteCount": 0,
          "id": 0,
          "title": "sad",
          "upvoteCount": 0,
        }
      }
      postId={0}
    />
  </View>
  <FlatList
    data={
      Array [
        Object {
          "children": Array [
            3,
            4,
            5,
          ],
          "content": "ccc1",
          "downvoteCount": 0,
          "id": 1,
          "upvoteCount": 0,
        },
        Object {
          "children": Array [],
          "content": "ccc2",
          "downvoteCount": 0,
          "id": 2,
          "upvoteCount": 0,
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Component: PostView renders with no comment 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 22,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "lightgrey",
        "padding": 10,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      name="close"
      onPress={[Function]}
      size={20}
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 25,
          "paddingTop": 10,
        }
      }
    >
       
      sad
       
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       
      ccc
       
    </Text>
    <Connect(OptionsRow)
      currentComment={
        Object {
          "children": Array [],
          "content": "ccc",
          "downvoteCount": 0,
          "id": 0,
          "title": "sad",
          "upvoteCount": 0,
        }
      }
      postId={0}
    />
  </View>
</View>
`;
